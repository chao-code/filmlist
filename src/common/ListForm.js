import { Component } from 'react'
import PropTypes from 'prop-types'
import update from 'immutability-helper'
import classNames from 'classnames'
import FilmSearch from './FilmSearch'
import Button from './Button'
import Poster from './Poster'
import TitleYear from './TitleYear'

const indexArray = n => Array(n).fill().map((_, i) => `${i + 1}`)

class ListForm extends Component {
  static propTypes = {
    list: PropTypes.shape({
      subject: PropTypes.string,
      listedBy: PropTypes.string,
      passcode: PropTypes.string,
      description: PropTypes.string,
      films: PropTypes.array,
    }),
    onSave: PropTypes.func.isRequired,
    onCancel: PropTypes.func.isRequired,
  }

  static defaultProps = {
    list: {
      subject: '',
      listedBy: '',
      passcode: '',
      description: '',
      films: [],
    },
  }

  state = {
    ...this.props.list,
    indexInputs: indexArray(this.props.list.films.length),
    showInvalid: false,
    alert: null,
  }

  onInputChange = event => {
    const { value, name } = event.target
    this.setState({
      [name]: value,
      alert: null,
    })
  }

  onIndexInputChange = index => event => {
    const { value } = event.target
    this.setState(prevState => ({
      indexInputs: update(prevState.indexInputs, {
        [index]: { $set: value },
      }),
      alert: Number.isInteger(+value) ? null : {
        type: 'warning',
        message: `"${value}" is not a valid index. Please enter an number then press Enter.`,
      },
    }))
  }

  onCommentInputChange = index => event => {
    const { value } = event.target
    this.setState(prevState => ({
      films: update(prevState.films, {
        [index]: {
          comment: { $set: value },
        },
      }),
    }))
  }

  dismissAlert = () => this.setState({ alert: null })

  addFilm = newFilm => this.setState(prevState => {
    const { films, indexInputs } = prevState
    if (films.find(film => film.imdbID === newFilm.imdbID)) {
      return {
        alert: {
          type: 'info',
          message: `${newFilm.title} (${newFilm.year}) is already in the list.`,
        },
      }
    }
    return {
      films: [...films, newFilm],
      indexInputs: [...indexInputs, `${indexInputs.length + 1}`],
      alert: null,
    }
  })

  removeFilm = index => () => this.setState(prevState => {
    const { films } = prevState
    return {
      films: [...films.slice(0, index), ...films.slice(index + 1)],
      indexInputs: indexArray(films.length - 1),
      alert: null,
    }
  })

  moveFilm = index => event => {
    if (event.key !== 'Enter') return
    const { value } = event.target
    let newIndex = value - 1
    if (newIndex === index) return
    if (Number.isNaN(newIndex)) {
      this.setState(prevState => {
        const { indexInputs, films } = prevState
        const { title, year } = films[index]
        return {
          indexInputs: update(indexInputs, {
            [index]: { $set: `${index + 1}` },
          }),
          alert: {
            type: 'danger',
            message: `Cannot move ${title} (${year}). "${value}" is not a valid index.`,
          },
        }
      })
    } else {
      this.setState(({ films }) => {
        const film = films[index]
        if (newIndex < 0) {
          newIndex = 0
        }
        if (newIndex >= films.length) {
          newIndex = films.length - 1
        }
        return {
          films: update(films, { $splice: [[index, 1], [newIndex, 0, film]] }),
          indexInputs: indexArray(films.length),
          alert: null,
        }
      })
    }
  }

  save = () => {
    const { showInvalid, alert, indexInputs, ...list } = this.state
    if (!list.films.length) {
      this.setState({
        showInvalid: true,
        alert: {
          type: 'danger',
          message: 'Cannot save list. Please add a film.',
        },
      })
    } else if (!list.subject || !list.listedBy) {
      this.setState({
        showInvalid: true,
        alert: {
          type: 'danger',
          message: 'Cannot save list. Please fill in all required fields.',
        },
      })
    } else {
      this.props.onSave(list)
    }
  }

  cancel = () => this.props.onCancel()

  renderListInfo() {
    const { subject, listedBy, passcode, description, showInvalid } = this.state
    return (
      <div className="row">
        <div className="col-md">
          <div className="form-group">
            <label className="form-control-label" htmlFor="subjectInput">Subject</label>
            <input
              className={classNames('form-control', { 'is-invalid': showInvalid && !subject })}
              type="text"
              id="subjectInput"
              name="subject"
              value={subject || ''}
              onChange={this.onInputChange}
            />
            <small className={showInvalid && !subject ? 'invalid-feedback' : 'form-text text-muted'}>
              Required
            </small>
          </div>
          <div className="form-group">
            <label className="form-control-label" htmlFor="listedByInput">Listed By</label>
            <input
              className={classNames('form-control', { 'is-invalid': showInvalid && !listedBy })}
              type="text"
              id="listedByInput"
              name="listedBy"
              value={listedBy || ''}
              onChange={this.onInputChange}
            />
            <small className={showInvalid && !listedBy ? 'invalid-feedback' : 'form-text text-muted'}>
              Required
            </small>
          </div>
          <div className="form-group">
            <label className="form-control-label" htmlFor="passcodeInput">Passcode</label>
            <input
              className="form-control"
              type="text"
              id="passcodeInput"
              name="passcode"
              value={passcode || ''}
              onChange={this.onInputChange}
            />
            <small className="form-text text-muted">
              Use Passcode to edit or delete this list
            </small>
          </div>
        </div>
        <div className="col-md">
          <div className="form-group">
            <label className="form-control-label" htmlFor="descriptionInput">Description</label>
            <textarea
              className="form-control"
              id="descriptionInput"
              rows="10"
              name="description"
              value={description || ''}
              onChange={this.onInputChange}
            />
          </div>
        </div>
      </div>
    )
  }

  renderFormActions() {
    const { showInvalid, films } = this.state
    return (
      <div className="row my-4">
        <div className="col col-md">
          <div className="input-group">
            <div className="input-group-prepend">
              <div className="input-group-text">Add to List</div>
            </div>
            <FilmSearch
              inputClass={classNames({ 'is-invalid': showInvalid && !films.length })}
              resultsContainerClass="film-results"
              placeholder="Enter Film Name..."
              onResultClick={this.addFilm}
            />
          </div>
        </div>
        <div className="col-5 col-md">
          <Button primary className="float-right" onClick={this.save}>Save</Button>
          <Button secondary className="float-right mr-3" onClick={this.cancel}>Cancel</Button>
        </div>
      </div>
    )
  }

  renderAlertMessage() {
    const { type, message } = this.state.alert
    return (
      <div className={`alert fade show alert-${type}`}>
        {message}
        <button type="button" className="close" aria-label="Close" onClick={this.dismissAlert}>
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
    )
  }

  renderListFilms() {
    const { films, indexInputs } = this.state
    return (
      <ul className="list-group">
        {films.map(({ title, year, poster, comment, imdbID }, index) => (
          <li className="list-group-item" key={imdbID}>
            <div className="row px-3">
              <div className="col-2 col-lg-1 px-1 px-md-2">
                <Poster poster={poster} />
              </div>
              <div className="col">
                <input
                  className="form-control text-center mr-2 py-1 px-1 mb-1 order-input"
                  type="text"
                  value={indexInputs[index]}
                  onChange={this.onIndexInputChange(index)}
                  onKeyPress={this.moveFilm(index)}
                />
                <h5 className="film-title">
                  <TitleYear title={title} year={year} />
                </h5>
                <input
                  className="form-control form-control-sm mt-3"
                  type="text"
                  placeholder="Add a Comment"
                  value={comment || ''}
                  onChange={this.onCommentInputChange(index)}
                />
              </div>
              <div className="col-1">
                <button className="close" onClick={this.removeFilm(index)}>
                  <span>&times;</span>
                </button>
              </div>
            </div>
          </li>
        ))}
      </ul>
    )
  }

  render() {
    return (
      <div>
        {this.renderListInfo()}
        {this.renderFormActions()}
        {this.state.alert &&
        this.renderAlertMessage()
        }
        {this.renderListFilms()}
      </div>
    )
  }
}

export default ListForm
