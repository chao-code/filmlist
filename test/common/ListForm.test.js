import { shallow } from 'enzyme'
import ListForm from '../../src/common/ListForm'
import { mockListDetail } from '../mocks/dataMock'

const { createdOn, filmCount, ...listDetail } = mockListDetail

describe('<ListForm />', () => {
  it('should shallow the empty form', () => {
    const props = {
      onSave: () => {},
      onCancel: () => {},
    }
    const wrapper = shallow(<ListForm {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should shallow the form filled with list', () => {
    const props = {
      list: listDetail,
      onSave: () => {},
      onCancel: () => {},
    }
    const wrapper = shallow(<ListForm {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should add a film to the list', () => {
    const props = {
      list: {
        ...listDetail,
        films: listDetail.films.slice(0, 2),
      },
      onSave: () => {},
      onCancel: () => {},
    }
    const wrapper = shallow(<ListForm {...props} />)
    expect(wrapper.find('li').length).toBe(2)
    wrapper.instance().addFilm(listDetail.films[2])
    wrapper.update()
    expect(wrapper.find('li').length).toBe(3)
    expect(wrapper.find('Poster').at(2).prop('poster')).toBe(listDetail.films[2].poster)
  })

  it('should change comment input', () => {
    const props = {
      list: listDetail,
      onSave: () => {},
      onCancel: () => {},
    }
    const event = { target: { value: 'a comment' } }
    const wrapper = shallow(<ListForm {...props} />)
    let input = wrapper.find('input[placeholder="Add a Comment"]').at(2)
    expect(input.prop('value')).toBe(listDetail.films[2].comment)
    input.simulate('change', event)
    input = wrapper.find('input[placeholder="Add a Comment"]').at(2)
    expect(input.prop('value')).toBe(event.target.value)
  })

  it('should change index input', () => {
    const props = {
      list: listDetail,
      onSave: () => {},
      onCancel: () => {},
    }
    const event = { target: { value: '2' } }
    const wrapper = shallow(<ListForm {...props} />)
    let input = wrapper.find('input.order-input').first()
    expect(input.prop('value')).toBe('1')
    input.simulate('change', event)
    input = wrapper.find('input.order-input').first()
    expect(input.prop('value')).toBe('2')
  })

  it('should move a film to the new index', () => {
    const props = {
      list: listDetail,
      onSave: () => {},
      onCancel: () => {},
    }
    const event = {
      key: 'Enter',
      target: { value: '1' },
    }
    const wrapper = shallow(<ListForm {...props} />)
    wrapper.find('input.order-input').at(2).simulate('keyPress', event)
    expect(wrapper.find('Poster').at(0).prop('poster')).toBe(listDetail.films[2].poster)
    expect(wrapper.find('Poster').at(1).prop('poster')).toBe(listDetail.films[0].poster)
    expect(wrapper.find('Poster').at(2).prop('poster')).toBe(listDetail.films[1].poster)
  })

  it('should remove a film from the list', () => {
    const props = {
      list: listDetail,
      onSave: () => {},
      onCancel: () => {},
    }
    const wrapper = shallow(<ListForm {...props} />)
    expect(wrapper.find('li').length).toBe(listDetail.films.length)
    wrapper.find('button.close').at(1).simulate('click')
    expect(wrapper.find('Poster').at(0).prop('poster')).toBe(listDetail.films[0].poster)
    expect(wrapper.find('Poster').at(1).prop('poster')).toBe(listDetail.films[2].poster)
  })

  it('should validate before save', () => {
    const onSave = jest.fn()
    const props = {
      onSave,
      onCancel: () => {},
    }
    const wrapper = shallow(<ListForm {...props} />)
    wrapper.find('Button').first().simulate('click')
    expect(wrapper.find('#subjectInput').hasClass('is-invalid')).toBeTruthy()
    expect(wrapper.find('#listedByInput').hasClass('is-invalid')).toBeTruthy()
    expect(wrapper.find('Connect(FilmSearch)').prop('inputClass')).toBe('is-invalid')
    expect(wrapper.find('.alert').length).toBe(1)
    expect(onSave.mock.calls.length).toBe(0)
  })

  it('should save the list', () => {
    const onSave = jest.fn()
    const props = {
      list: listDetail,
      onSave,
      onCancel: () => {},
    }
    const wrapper = shallow(<ListForm {...props} />)
    wrapper.find('Button').first().simulate('click')
    expect(onSave.mock.calls[0][0]).toEqual(listDetail)
  })

  it('should cancel editing the list', () => {
    const onCancel = jest.fn()
    const props = {
      list: listDetail,
      onSave: () => {},
      onCancel,
    }
    const wrapper = shallow(<ListForm {...props} />)
    wrapper.find('Button').at(1).simulate('click')
    expect(onCancel.mock.calls.length).toBe(1)
  })
})
