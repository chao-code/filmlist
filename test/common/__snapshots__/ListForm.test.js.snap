// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListForm /> should shallow the empty form 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-md"
    >
      <div
        className="form-group"
      >
        <label
          className="form-control-label"
          htmlFor="subjectInput"
        >
          Subject
        </label>
        <input
          className="form-control"
          id="subjectInput"
          name="subject"
          onChange={[Function]}
          type="text"
          value=""
        />
        <small
          className="form-text text-muted"
        >
          Required
        </small>
      </div>
      <div
        className="form-group"
      >
        <label
          className="form-control-label"
          htmlFor="listedByInput"
        >
          Listed By
        </label>
        <input
          className="form-control"
          id="listedByInput"
          name="listedBy"
          onChange={[Function]}
          type="text"
          value=""
        />
        <small
          className="form-text text-muted"
        >
          Required
        </small>
      </div>
      <div
        className="form-group"
      >
        <label
          className="form-control-label"
          htmlFor="passcodeInput"
        >
          Passcode
        </label>
        <input
          className="form-control"
          id="passcodeInput"
          name="passcode"
          onChange={[Function]}
          type="text"
          value=""
        />
        <small
          className="form-text text-muted"
        >
          Use Passcode to edit or delete this list
        </small>
      </div>
    </div>
    <div
      className="col-md"
    >
      <div
        className="form-group"
      >
        <label
          className="form-control-label"
          htmlFor="descriptionInput"
        >
          Description
        </label>
        <textarea
          className="form-control"
          id="descriptionInput"
          name="description"
          onChange={[Function]}
          rows="10"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row my-4"
  >
    <div
      className="col col-md"
    >
      <div
        className="input-group"
      >
        <div
          className="input-group-prepend"
        >
          <div
            className="input-group-text"
          >
            Add to List
          </div>
        </div>
        <Connect(FilmSearch)
          inputClass=""
          onResultClick={[Function]}
          placeholder="Enter Film Name..."
          resultsContainerClass="film-results"
        />
      </div>
    </div>
    <div
      className="col-5 col-md"
    >
      <Button
        className="float-right"
        danger={false}
        href=""
        link={false}
        onClick={[Function]}
        primary={true}
        secondary={false}
        small={false}
      >
        Save
      </Button>
      <Button
        className="float-right mr-3"
        danger={false}
        href=""
        link={false}
        onClick={[Function]}
        primary={false}
        secondary={true}
        small={false}
      >
        Cancel
      </Button>
    </div>
  </div>
  <ul
    className="list-group"
  />
</div>
`;

exports[`<ListForm /> should shallow the form filled with list 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-md"
    >
      <div
        className="form-group"
      >
        <label
          className="form-control-label"
          htmlFor="subjectInput"
        >
          Subject
        </label>
        <input
          className="form-control"
          id="subjectInput"
          name="subject"
          onChange={[Function]}
          type="text"
          value="subject"
        />
        <small
          className="form-text text-muted"
        >
          Required
        </small>
      </div>
      <div
        className="form-group"
      >
        <label
          className="form-control-label"
          htmlFor="listedByInput"
        >
          Listed By
        </label>
        <input
          className="form-control"
          id="listedByInput"
          name="listedBy"
          onChange={[Function]}
          type="text"
          value="listed by"
        />
        <small
          className="form-text text-muted"
        >
          Required
        </small>
      </div>
      <div
        className="form-group"
      >
        <label
          className="form-control-label"
          htmlFor="passcodeInput"
        >
          Passcode
        </label>
        <input
          className="form-control"
          id="passcodeInput"
          name="passcode"
          onChange={[Function]}
          type="text"
          value="passcode"
        />
        <small
          className="form-text text-muted"
        >
          Use Passcode to edit or delete this list
        </small>
      </div>
    </div>
    <div
      className="col-md"
    >
      <div
        className="form-group"
      >
        <label
          className="form-control-label"
          htmlFor="descriptionInput"
        >
          Description
        </label>
        <textarea
          className="form-control"
          id="descriptionInput"
          name="description"
          onChange={[Function]}
          rows="10"
          value="description"
        />
      </div>
    </div>
  </div>
  <div
    className="row my-4"
  >
    <div
      className="col col-md"
    >
      <div
        className="input-group"
      >
        <div
          className="input-group-prepend"
        >
          <div
            className="input-group-text"
          >
            Add to List
          </div>
        </div>
        <Connect(FilmSearch)
          inputClass=""
          onResultClick={[Function]}
          placeholder="Enter Film Name..."
          resultsContainerClass="film-results"
        />
      </div>
    </div>
    <div
      className="col-5 col-md"
    >
      <Button
        className="float-right"
        danger={false}
        href=""
        link={false}
        onClick={[Function]}
        primary={true}
        secondary={false}
        small={false}
      >
        Save
      </Button>
      <Button
        className="float-right mr-3"
        danger={false}
        href=""
        link={false}
        onClick={[Function]}
        primary={false}
        secondary={true}
        small={false}
      >
        Cancel
      </Button>
    </div>
  </div>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item"
      key="tt0123456"
    >
      <div
        className="row px-3"
      >
        <div
          className="col-2 col-lg-1 px-1 px-md-2"
        >
          <Poster
            crop={false}
            poster="poster1.jpg"
          />
        </div>
        <div
          className="col"
        >
          <input
            className="form-control text-center mr-2 py-1 px-1 mb-1 order-input"
            onChange={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="1"
          />
          <h5
            className="film-title"
          >
            <TitleYear
              title="Film A"
              year="2000"
            />
          </h5>
          <input
            className="form-control form-control-sm mt-3"
            onChange={[Function]}
            placeholder="Add a Comment"
            type="text"
            value=""
          />
        </div>
        <div
          className="col-1"
        >
          <button
            className="close"
            onClick={[Function]}
          >
            <span>
              ×
            </span>
          </button>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
      key="tt1234567"
    >
      <div
        className="row px-3"
      >
        <div
          className="col-2 col-lg-1 px-1 px-md-2"
        >
          <Poster
            crop={false}
            poster="poster2.jpg"
          />
        </div>
        <div
          className="col"
        >
          <input
            className="form-control text-center mr-2 py-1 px-1 mb-1 order-input"
            onChange={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="2"
          />
          <h5
            className="film-title"
          >
            <TitleYear
              title="Film B"
              year="2010"
            />
          </h5>
          <input
            className="form-control form-control-sm mt-3"
            onChange={[Function]}
            placeholder="Add a Comment"
            type="text"
            value=""
          />
        </div>
        <div
          className="col-1"
        >
          <button
            className="close"
            onClick={[Function]}
          >
            <span>
              ×
            </span>
          </button>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
      key="tt2345678"
    >
      <div
        className="row px-3"
      >
        <div
          className="col-2 col-lg-1 px-1 px-md-2"
        >
          <Poster
            crop={false}
            poster="poster3.jpg"
          />
        </div>
        <div
          className="col"
        >
          <input
            className="form-control text-center mr-2 py-1 px-1 mb-1 order-input"
            onChange={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="3"
          />
          <h5
            className="film-title"
          >
            <TitleYear
              title="Film C"
              year="2015"
            />
          </h5>
          <input
            className="form-control form-control-sm mt-3"
            onChange={[Function]}
            placeholder="Add a Comment"
            type="text"
            value="no comment"
          />
        </div>
        <div
          className="col-1"
        >
          <button
            className="close"
            onClick={[Function]}
          >
            <span>
              ×
            </span>
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
